test/unit/common/test_shardtrie.py

```diff
@@ -0,0 +1,312 @@
+            trie.get_node('abc')
+
+    def test_shardtrie_join(self):
+        """Test joining trie"""
+        trie = self.trie
+
+        subtrie = trie.split_trie('ab')
+        self.assertEqual(len(list(trie.get_data_nodes())), 1)
+
+        trie.join_subtrie(subtrie)
+        self.assertEqual(len(list(trie.get_data_nodes())), 3)
+
+    def test_shardtrie_join_subtrie_not_trie(self):
+        """Test joining trie when subtrie is not an instance of ShardTrie"""
+        trie = self.trie
+        subtrie = trie.split_trie('ab')
+
+        with self.assertRaises(shardtrie.ShardTrieException):
+            trie.join_subtrie({'this_is': 'invalid'})
+
+    def test_shardtrie_popping_iterator(self):
+        """Test iterating across the trie while removing visited elements"""
+        trie = self.trie
+
+        list(trie.root.popping_iter())
+        # We cannot use the following line as is_empty checks that the key
+        #  is empty string, popping_iterator does not set the key to ''
+        # self.assertTrue(trie.is_empty())
+        self.assertEqual(trie.root.children, {})
+
+    def test_shardtrie_delete(self):
+        """Test deleting a node from shardtrie"""
+        trie = self.trie
+        number_data_nodes = trie.root.count_data_nodes()
+
+        # Delete a non data node
+        self.assertTrue(trie.delete('ab'))
+        self.assertEqual(trie.root.count_data_nodes(), number_data_nodes)
+
+        # Delete a data node
+        self.assertTrue(trie.delete('abc'))
+        self.assertEqual(trie.root.count_data_nodes(), number_data_nodes - 1)
+
+    def test_shardtrie_add(self):
+        """Test adding a node to the shardtrie"""
+        trie = self.trie
+        number_data_nodes = trie.root.count_data_nodes()
+
+        self.assertTrue(trie.add('abce'))
+        self.assertEqual(trie.root.count_data_nodes(), number_data_nodes + 1)
+
+    def test_shardtrie_add_distributed(self):
+        """Test adding a node to a distributed shardtrie"""
+        trie = self.trie
+
+        trie.split_trie('ab')
+        with self.assertRaises(shardtrie.ShardTrieDistributedBranchException):
+            trie.add('abce')
+
+    def test_shardtrie_get_important_nodes(self):
+        """Test iterating across data nodes"""
+        trie = self.trie
+
+        # Iterate over all data nodes
+        self.assertTrue(
+            all([x.data != None for x in trie.get_important_nodes()]))
+
+        # Iterate over only 2 data nodes
+        self.assertEqual(
+            len(list(trie.get_important_nodes(limit = 2))),
+            2)
+
+        # Iterate starting after node abc
+        marker = trie.get_node('abc')
+        self.assertEqual(
+            list(trie.get_important_nodes(marker=marker))[0].full_key(),
+            'abd')
+
+        # Iterate starting at 'abc'
+        self.assertEqual(
+            list(trie.get_important_nodes('abc'))[0].full_key(),
+            'abc')
```

Comment:
Also, I added the limit stuff here cause originally I thought this would be where we'd use it. In the current interation (i'd have to check), because on container GET we grab CONTAINER_LISTING_LIMIT at a time, I think we are letting the container backend class deal with the limits for us, and we just keep track of how many are left.

TL;DR: I'm not sure we are actually using the limit stuff in shardtrie class... although maybe we should be.. just making a note cause one day we may decide to remove it on a clean up. But we'll decide on that later. :)